ttnn/cpp/ttnn/operations/eltwise/binary_ng/device/binary_ng_program_factory.cpp

```diff
@@ -0,0 +1,397 @@
+
+    uint32_t a_single_tile_size = tt_metal::detail::TileSize(a_data_format);
+    uint32_t b_single_tile_size = tt_metal::detail::TileSize(b_data_format);
+    uint32_t c_single_tile_size = tt_metal::detail::TileSize(c_data_format);
+
+    uint32_t num_output_tiles = c.volume() / c.tensor_spec().tile().get_tile_hw();
+
+    // we parallelize the computation across the output tiles
+    constexpr bool row_major = true;
+    auto compute_with_storage_grid_size = device->compute_with_storage_grid_size();
+    uint32_t num_cores_x = compute_with_storage_grid_size.x;
+    uint32_t num_cores_y = compute_with_storage_grid_size.y;
+    auto all_device_cores = CoreRange({0, 0}, {num_cores_x - 1, num_cores_y - 1});
+
+    Buffer* a_buffer = a.buffer();
+    Buffer* b_buffer = nullptr;
+    Buffer* c_buffer = c.buffer();
+
+    // How many tiles to store per input CB (double buffer)
+    constexpr uint32_t num_tiles_per_cb = 2;
+    auto [a_cb, a_cb_handle] =
+        create_cb(tt::CB::c_in0, program, all_device_cores, a_single_tile_size, num_tiles_per_cb, a_data_format);
+    auto [c_cb, c_cb_handle] =
+        create_cb(tt::CB::c_out0, program, all_device_cores, c_single_tile_size, num_tiles_per_cb, c_data_format);
+
+    // If b is a scalar, we only need one tile in the CB
+    uint32_t b_num_tiles_per_cb = b_buffer != nullptr ? num_tiles_per_cb : 1;
+    auto [b_cb, b_cb_handle] =
+        create_cb(tt::CB::c_in1, program, all_device_cores, b_single_tile_size, b_num_tiles_per_cb, b_data_format);
+
+    auto a_is_dram = static_cast<uint32_t>(a_buffer->buffer_type() == tt_metal::BufferType::DRAM);
+    bool b_is_dram = false;
+    auto c_is_dram = static_cast<uint32_t>(c_buffer->buffer_type() == tt_metal::BufferType::DRAM);
+
+    auto kernel_config = CMAKE_UNIQUE_NAMESPACE::BinaryNgKernelConfig(operation_attributes.subtile_broadcast_type);
+
+    // READER KERNEL
+    auto reader_kernel_id = tt_metal::CreateKernel(
+        program,
+        get_kernel_file_path(kernel_config.reader_kernel),
+        all_device_cores,
+        tt_metal::ReaderDataMovementConfig({a_is_dram}));
+
+    // WRITER KERNEL
+    auto writer_kernel = CMAKE_UNIQUE_NAMESPACE::KernelName::WriterScalar;
+    auto compute_kernel = CMAKE_UNIQUE_NAMESPACE::KernelName::ComputeScalar;
+    if (b.has_value()) {
+        b_buffer = b->buffer();
+        b_is_dram = static_cast<uint32_t>(b_buffer->buffer_type() == tt_metal::BufferType::DRAM);
+        writer_kernel = kernel_config.writer_kernel;
+        compute_kernel = kernel_config.compute_kernel;
+    }
+
+    auto writer_kernel_id = tt_metal::CreateKernel(
+        program,
+        get_kernel_file_path(writer_kernel),
+        all_device_cores,
+        tt_metal::WriterDataMovementConfig({b_is_dram, c_is_dram}));
+
+    // COMPUTE KERNEL
+    bool fp32_dest_acc_en = c_data_format == tt::DataFormat::UInt32 || c_data_format == tt::DataFormat::Int32 ||
+                            c_data_format == tt::DataFormat::Float32;
```

Comment:
Not important for now, but is this something we want to expose later on?

Reply:
Yes, this is really more of a placeholder for now as a reminder to support more dtypes.